### PR TITLE
[NEAT-419, NEAT-417] 🐋 Cleanup serialization Part 2

### PR DESCRIPTION
**Pull request by doctrino**
Replace all rules `Serializers` and instead use a built in mechanism in `pydantic` called `context` to control how reference should be set in when dumping rules to `dict/json` as well as whether to remove default prefix,space, and version.

This has the benefit of reducing the code base as well as moving the serialization to the `DMSProperty`, `InformationProperty, `InformationClass`, making it clearer for the developer how serialization is done.

### CODE REVIEW

File: cognite/neat/rules/models/information/_rules.py [hunk 1]
```diff
@@ -171,15 +205,13 @@ class InformationProperty(SheetRow):
     comment: str | None = Field(alias="Comment", default=None)
     inherited: bool = Field(
         default=False,
+        exclude=True,
```
Comment (doctrino) [inline cognite/neat/rules/models/information/_rules.py:208]: This fixes the bug found by Kristian NEAT-417

File: cognite/neat/rules/models/_base_rules.py [hunk 2]
```diff
@@ -210,10 +190,70 @@ def headers_by_sheet(cls, by_alias: bool = False) -> dict[str, list[str]]:
...
+        entities_exclude_defaults: bool = True,
+    ) -> dict[str, Any]:
+        """Dump the model to a dictionary.
+
+        This is used in the Exporters to dump rules in the required format.
+        """
+        for field_name in self.model_fields.keys():
+            value = getattr(self, field_name)
+            # Ensure deterministic order of properties
+            if isinstance(value, SheetList):
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_base_rules.py:215]: do we have fields which are not of type `SheetList`

File: cognite/neat/rules/models/_base_rules.py [hunk 3]
```diff
@@ -210,10 +190,70 @@ def headers_by_sheet(cls, by_alias: bool = False) -> dict[str, list[str]]:
...
+        mode: Literal["python", "json"] = "python",
+        by_alias: bool = False,
+        exclude: IncEx = None,
+        exclude_none: bool = False,
+        exclude_unset: bool = False,
+        exclude_defaults: bool = False,
+        as_reference: bool = False,
+        entities_exclude_defaults: bool = True,
+    ) -> dict[str, Any]:
+        """Dump the model to a dictionary.
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_base_rules.py:208]: A bit more context would be helpful especially all these input arg need explanation of what they are

File: cognite/neat/rules/models/_base_rules.py [hunk 4]
```diff
@@ -210,10 +190,70 @@ def headers_by_sheet(cls, by_alias: bool = False) -> dict[str, list[str]]:
...
+        output = self.model_dump(
+            mode=mode,
+            by_alias=by_alias,
+            exclude=exclude_input,
+            exclude_none=exclude_none,
+            exclude_unset=exclude_unset,
+            exclude_defaults=exclude_defaults,
+            context=context,
+        )
+        if self.reference is not None and not (isinstance(exclude, dict | set) and "reference" in exclude):
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_base_rules.py:244]: A bit confusing if statement, 
I would expect this `if` statement to look more like:

`if self.reference and exclude and "reference" not in exclude:`

File: cognite/neat/rules/models/_base_rules.py [hunk 5]
```diff
@@ -210,10 +190,70 @@ def headers_by_sheet(cls, by_alias: bool = False) -> dict[str, list[str]]:
...
+            value = getattr(self, field_name)
+            # Ensure deterministic order of properties
+            if isinstance(value, SheetList):
+                value.sort(key=lambda x: x._identifier())
+
+        context: dict[str, Any] = {"as_reference": as_reference}
+        if entities_exclude_defaults:
+            context["metadata"] = self.metadata
+
+        if self.reference is not None:
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_base_rules.py:222]:
```suggestion
        if self.reference:
```

File: cognite/neat/rules/models/_base_rules.py [hunk 6]
```diff
@@ -210,10 +190,70 @@ def headers_by_sheet(cls, by_alias: bool = False) -> dict[str, list[str]]:
...
+            exclude_input: IncEx
+            if isinstance(exclude, dict):
```
Comment (nikokaoja) [inline cognite/neat/rules/models/_base_rules.py:224]: why `exclude` can be of various types? some explanation would be helpful as a comment or similar

File: cognite/neat/rules/models/information/_rules.py (hunk 1 above)
Comment (nikokaoja) [reply cognite/neat/rules/models/information/_rules.py:208]: This is to exclude when exporting rules

File: cognite/neat/rules/models/information/_rules.py [hunk 7]
```diff
@@ -171,15 +205,13 @@ class InformationProperty(SheetRow):
...
         alias="Inherited",
         description="Flag to indicate if the property is inherited, only use for internal purposes",
     )
 
-    @field_serializer("max_count", when_used="json-unless-none")
-    def serialize_max_count(self, value: int | float | None) -> int | float | None | str:
-        if isinstance(value, float) and math.isinf(value):
-            return None
-        return value
+    def _identifier(self) -> tuple[Hashable, ...]:
```
Comment (nikokaoja) [inline cognite/neat/rules/models/information/_rules.py:213]: what is intended used for this method?

File: cognite/neat/rules/models/information/_rules.py [hunk 8]
```diff
@@ -233,6 +265,30 @@ def set_type_for_default(self):
...
 
+    @field_serializer("max_count", when_used="json-unless-none")
+    def serialize_max_count(self, value: int | float | None) -> int | float | None | str:
+        if isinstance(value, float) and math.isinf(value):
+            return None
+        return value
+
+    @field_serializer("reference", when_used="always")
+    def set_reference(self, value: Any, info: SerializationInfo) -> str | None:
+        if isinstance(info.context, dict) and info.context.get("as_reference") is True:
```
Comment (nikokaoja) [inline cognite/neat/rules/models/information/_rules.py:276]: do you set `as_reference` when you load rules from somewhere?

File: cognite/neat/rules/models/information/_rules.py [hunk 9]
```diff
@@ -233,6 +265,30 @@ def set_type_for_default(self):
...
+                    prefix=str(self.class_.prefix),
+                    suffix=self.class_.suffix,
+                    property=self.property_,
+                )
+            )
+        return str(value) if value is not None else None
+
+    @field_serializer("class_", "value_type", when_used="unless-none")
+    def remove_default_prefix(self, value: Any, info: SerializationInfo) -> str:
+        if (metadata := _metadata(info.context)) and isinstance(value, Entity):
```
Comment (nikokaoja) [inline cognite/neat/rules/models/information/_rules.py:288]: where `_metadata` is now coming from ?

**Review by nikokaoja — APPROVED** [review]
I am a bit worried about `rules.models` I think it became complex with pydantic low level feature implementations. It is trade-off between writing custom, yet probably simple methods, vs implementing something from pydantic  low-level functionality. 

I will approve your PRs, but context on methods either through docstring or comments in the code base are essential to understand flow with an ease.

File: cognite/neat/rules/models/_base_rules.py (hunk 2 above)
Comment (doctrino) [reply cognite/neat/rules/models/_base_rules.py:215]: Yes Metadata.

File: cognite/neat/rules/models/_base_rules.py (hunk 6 above)
Comment (doctrino) [reply cognite/neat/rules/models/_base_rules.py:224]: This is what pydantic accepts. Adding comment to explain why this is done.

File: cognite/neat/rules/models/information/_rules.py (hunk 9 above)
Comment (doctrino) [reply cognite/neat/rules/models/information/_rules.py:288]: It is a small helper function

File: cognite/neat/rules/models/information/_rules.py (hunk 8 above)
Comment (doctrino) [reply cognite/neat/rules/models/information/_rules.py:276]: This is not for loading, it is for serialization. added an extra comment

**Pull request merged**